sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -90,6 +90,34 @@ case class ClusteredDistribution(
   }
 }
 
+/**
+ * Represents the requirement of distribution on the stateful operator.
+ *
+ * Each partition in stateful operator initializes state store(s), which are independent with state
+ * store(s) in other partitions. Since it is not possible to repartition the data in state store,
+ * Spark should make sure the physical partitioning of the stateful operator is unchanged across
+ * Spark versions. Violation of this requirement may bring silent correctness issue.
+ *
+ * Since this distribution relies on [[HashPartitioning]] on the physical partitioning of the
+ * stateful operator, only [[HashPartitioning]] can satisfy this distribution.
+ */
+case class StatefulOpClusteredDistribution(
```

Comment:
I like the new name :) thanks for making it more specific.

Do we also need to update `HashShuffleSpec` so that two `HashPartitioning`s can be compatible with each other when checking against `StatefulOpClusteredDistribution`s? this is the previous behavior where Spark would avoid shuffle if both sides of the streaming join are co-partitioned.

Reply:
> Do we also need to update HashShuffleSpec so that two HashPartitionings can be compatible with each other when checking against StatefulOpClusteredDistributions? this is the previous behavior where Spark would avoid shuffle if both sides of the streaming join are co-partitioned.

Each input must follow the required distribution provided from stateful operator to respect the requirement of state partitioning. state partitioning is the first class, so even both sides of the streaming join are co-partitioned, Spark must perform shuffle if they don't match with state partitioning. (If that was the previous behavior, we broke something.)

Reply:
Actually, I think this PR **will skip shuffle** if both sides of a streaming join are co-partitioned. In `EnsureRequirements`, we currently mainly do two things:
1. check if output partitioning can satisfy the required distribution
2. if there are two children, check if they are compatible with each other, and insert shuffle if not.

In the step 2) we'd only consider `ClusteredDistribution` at the moment, so in case of `StatefulOpClusteredDistributions` this step is simply skipped. Consequently, Spark will skip shuffle even if only step 1) is successful. 

> State partitioning is the first class, so even both sides of the streaming join are co-partitioned, Spark must perform shuffle if they don't match with state partitioning.

I'm not quite sure about this. Shouldn't we retain the behavior before https://github.com/apache/spark/pull/32875? Quoting [the comment](https://github.com/apache/spark/pull/32875#issuecomment-1023829161) from @cloud-fan:

> I think this is kind of a potential bug. Let's say that we have 2 tables that can report hash partitioning optionally (e.g. controlled by a flag). Assume a streaming query is first run with the flag off, which means the tables do not report hash partitioning, then Spark will add shuffles before the stream-stream join, and the join state (steaming checkpoint) is partitioned by Spark's murmur3 hash function. Then we restart the streaming query with the flag on, and the 2 tables report hash partitioning (not the same as Spark's murmur3). Spark will not add shuffles before stream-stream join this time, and leads to wrong result, because the left/right join child is not co-partitioned with the join state in the previous run.

If we respect co-partitioning and avoid shuffle before https://github.com/apache/spark/pull/32875 but start shuffle after this PR, I think similar issue like described in the comment can happen?

Reply:
> 1. check if output partitioning can satisfy the required distribution

For stream-stream join, once each input satisfy the required "hash" distribution of each, they will be co-partitioned. stream-stream join must guarantee this.

Reply:
The problem brought up because ClusteredDistribution has much more relaxed requirement; what we really need to require is that for all children a specific tuple having specific grouping key must be bound to the deterministic partition "ID", which only HashClusteredDistribution could guarantee.

Reply:
I think one behavior difference between this PR and the state before https://github.com/apache/spark/pull/32875 is that, previously, we'd also check `spark.sql.shuffle.partitions` and insert shuffle if there's not enough parallelism from the input. However, this PR doesn't do that since it skips the step 2) above.

Reply:
HashClusteredDistribution also has a requirement of the number of partitions, step 1) should fulfill it.

Reply:
Yes, `StatefulOpClusteredDistributions` is very strict and requires numPartitions as well. I don't think we need extra co-partition check for it.

Reply:
I see, it should be good then!